core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/SkewJoin.java

```diff
@@ -0,0 +1,309 @@
+    }
+
+    private UnnestNode unnestBuildSide(TableScanNode build, SymbolReference skewedRightKey, List<Expression> inSkewedValues, Symbol partSymbol, Context context)
+    {
+        PlanNodeIdAllocator idAllocator = context.getIdAllocator();
+        Symbol partitionerSmbol = context.getSymbolAllocator().newSymbol("skewPartitioner", new ArrayType(IntegerType.INTEGER));
+
+        Expression ifSkewedKey = ifContains(skewedRightKey, inSkewedValues, ARRAY_PARTITIONER, DEFAULT_UNNEST_VALUE);
+
+        Assignments assignments = Assignments.builder()
+                .putIdentities(build.getOutputSymbols())
+                .put(partitionerSmbol, ifSkewedKey)
+                .build();
+
+        ProjectNode setPartitionArray = new ProjectNode(idAllocator.getNextId(), build, assignments);
+
+        return new UnnestNode(
+                idAllocator.getNextId(),
+                setPartitionArray,
+                build.getOutputSymbols(),
+                ImmutableList.of(new UnnestNode.Mapping(partitionerSmbol, ImmutableList.of(partSymbol))),
+                Optional.empty(),
+                JoinNode.Type.INNER,
+                Optional.empty());
+    }
+
+    private JoinNode rewriteJoinClause(JoinNode node, ProjectNode leftSource, Symbol leftKey, UnnestNode rightSource, Symbol rightKey, Context context)
+    {
+        List<EquiJoinClause> criteria = Stream
+                .concat(node.getCriteria().stream(), Stream.of(new JoinNode.EquiJoinClause(leftKey, rightKey)))
+                .collect(Collectors.toUnmodifiableList());
+
+        return new JoinNode(
+                context.getIdAllocator().getNextId(),
+                node.getType(),
+                leftSource,
+                rightSource,
+                criteria,
+                node.getLeftOutputSymbols(),
+                node.getRightOutputSymbols(),
+                node.isMaySkipOutputDuplicates(),
+                node.getFilter(),
+                node.getLeftHashSymbol(),
+                node.getRightHashSymbol(),
+                node.getDistributionType(),
+                node.isSpillable(),
+                node.getDynamicFilters(),
+                node.getReorderJoinStatsAndCost());
+    }
+
+    private ProjectNode dropAddedSymbols(JoinNode node, ImmutableSet<Symbol> symbolsToRemove, Context context)
+    {
+        ImmutableList<Symbol> newOutputSymbols = node.getOutputSymbols().stream()
+                .filter(symbol -> !symbolsToRemove.contains(symbol))
+                .collect(toImmutableList());
+
+        Assignments assignments = Assignments.builder().putIdentities(newOutputSymbols).build();
+
+        return new ProjectNode(context.getIdAllocator().getNextId(), node, assignments);
+    }
+
+    @Override
+    public Result apply(JoinNode joinNode, Captures captures, Context context)
+    {
+        if (joinNode.getDistributionType().isEmpty() || joinNode.getDistributionType().get() != JoinNode.DistributionType.PARTITIONED) {
```

Comment:
Basically add a comment that for Broadcast join it need not be applied as join is as distributed as splits.